### DISCUSSION TITLE
Adding Type Hints

### DISCUSSION THREAD

**Discussion by dvoraen** [evennia/evennia#2973] [Evennia Engine Development]
This is as much a set of queries as much as it is a project I kind of want to undertake in my spare time.  As you doubtless guessed from the subject, I'd like to see these added to Evennia, with an eye for Python 3.11's additions to the typing library (looking at you 'Self').  But like any project idea of mine, I have some questions. :)

- **Which branch would be most ideal to use to begin work: master (v0.95) or develop (targeting v1.0)?**
I feel like the latter would be the more ideal choice, due to the quantity of changes and additions, but I wanted to hear it straight from the horse's mouth as it were.
- **Should type hints be injected directly into the code, or should they instead be released as stubs?**
With my desire to have the 3.11 features included, I'm leaning towards the latter as the starting point, as that would be less invasive and would not create a version dependency that overrides what master/develop have.  It would require subsequent maintenance on the stubs for additions made to Evennia, true, but again I wanted to hear thoughts on this as to which would be deemed preferable.
- **How extensive should the hinting be?**
My intention is not to go postal in the code itself and type every single variable (I'm being a little hyperbolic here), but rather the essentials.  As such, my proposal (such as it is) is along these lines:
  - All function and method arguments and return types, where able.  This is the absolute minimum of the project.
  - All class variables and instance variables.
  - All global variables, where able.
  This particular point is where it becomes tricky.  For example, should `x = 'Hello'` be typed as `x: str = 'Hello'` or `x: Literal['Hello']` or just left alone since the type is understood?
  - Various protocols for less concrete arguments and return types.
  I bring these up as a separate point because, e.g., a Command's caller attribute can be one of multiple types, depending on the circumstances*, so it seems like a Protocol would be more appropriate to describe a caller rather than a concrete type.  Random aside, but I foresee evennia.typing.protocols being a thing under this project.

That's all I have for now.  I'm sure I'll think of something more later, while I'm at work, as usual. :)  I appreciate any time given to responses.

\* I'm operating off of v0.95 with this statement.  If v1.0's development has changed this, I'm just simply unaware and intend to include that in the research.

**Comment by Griatch** [comment]
Hi @dvoraen! 

We haven't focused on type hinting since, well, Evennia predates this being available in Python. Instead we have used Google docstrings to denote types (of course, this is for human eyes and not something a typechecker understands). If it's something you want to take on, that's awesome. Note that Evennia has not been "officially" tested with Python 3.11 at this time.

- If wanting to add this, it would need to be done against `develop` branch (1.0-dev). There will be no further features added to 0.9.5.
- Concerning stubs or inline: 
   - Python typing as a concept is still immature IMO, and there's still a lot of new features coming out. We could not support hints from 3.11 in 3.9 code (which dev also supports today). Inline typing also makes the code longer and harder to read overall. 
   - So it seems that stubs (or, better, a stub library) is the most flexible and non-invasive solution right now.
- The main reason for not 'going overboard' with hinting is because doing so makes the code harder to read. In a stub this matters less. I think handling functions/method args and their returns would be the main focus though. 
- Things like trying to typehint command's `caller` is not a high prio; there is a new Session-structure coming post 1.0 that would change that anyway.
- Evennia's existing docstring specify types, but often in a less specific way than is possible with typehints (so `list` rather than `List[str]` to denote a list of strings), but nevertheless, the docstrings should be used as a reference.

**Reply by dvoraen** [reply]
@Griatch, hi!  Thank you for replying.

A type stub library was the intention if that was the preferred route.  I'm dithering a little on exactly how to execute that.  My initial impression is to put the stubs off in their own kittycorner rather than put the stubs alongside the .py files.  That keeps things cleaner, I think, but would put the onus on Evennia coders for telling their type checkers and other tools where to find them.  Since stubs aren't mandatory for doing any work with Evennia whatsoever, I don't think this is really an issue.

I was severely considering trying to target Python 3.9 with the typing, but I think the benefits of ParamSpec, Self, and Variadic Generics (TypeVarTuple) are simply too beneficial to pass up.  I'm slowly conducting research as to whether it'd be possible to go with 3.9, with some adjustments.  Not knowing about Evennia's callbacks and decorators in detail is working against me here, and would be the primary impetus (via ParamSpec) for targeting 3.10 at the minimum.  I'm also still a little unclear on the role of the typing_extensions module with respect to everyday users, which would allow some "backporting" to previous versions.  I don't think it's part of the standard library(?), but again, I'm unclear on this point and need to look into it more.

I presume any and all existing docstrings for classes, functions, etc. would by necessity need to be included in any submitted stubs.  This just seems like an Occam's Razor type of subject to me, but I bring it up in case something didn't occur to me.

Either way, I have a starting point.  Now I just need more hours in the day. :)

**Reply by Griatch** [reply]
- I think it's fine to aim for 3.11 compliance for a stub file; I suspect before those stubs are a commonly used/referenced part of the library, we will likely have moved up to 3.10+ requirements at least (for example things like `case` is not possible while supporting 3.9). 
- I think it sounds good to make a stub package in their own 'kittycorner'. Putting them next to the `.py` files just makes it messier to find stuff.
- I'm not familiar enough with stubs to know if the docstrings _have_ to be included, but if they don't, I think they should not be - having docstrings in two places are just a recipe for making things very hard to maintain. Some of our docstrings are _very_ involved.

**Comment by dvoraen** [comment]
@Griatch I started this project by going through evennia.utils and a couple major questions arose.

### Disclaimer
I realize these questions probably have the following answer: "if you do, put them in their own PR and not as a bundle so any changes can be reviewed separately, CI, etc."

### Would you prefer I not make any code changes and keep it strictly to adding type stubs?
There were a couple places in utils/utils.py that gave me pause, because I was reading the code to make sure I understood the logical paths for any and all return values.  This may be a difference in style between us or is from code before you took over as a maintainer, but one example is as follows:
https://github.com/evennia/evennia/blob/7fff93647bad52210e7c976acea2921c2a8f4d98/evennia/utils/utils.py#L84

I know order of operators will take hold here (I had to refresh my memory on this), but it strikes me as difficult to read and requires a bit of thought to parse mentally.  I am uncertain what's gained by doing boolean shenanigans over a simple conditional return.  If `is_iter()` returns False, the first half of the `and` returns True, then moves on to evaluate and return `[obj]` because that's always True as it's not an empty list.  If `is_iter()` returns True, you get `obj` because `not is_iter() and [obj]` evalutes to False, which then chains into `False or obj` which returns `obj`.

Therefore, should it preferably look more like this for clarity's sake?
```python
return (not is_iter(obj) and [obj]) or obj
```
Personally, and this is where my sense of style difference comes in that we might disagree on, I would like to change it to this:
```python
return [obj] if not is_iter(obj) else obj
```

This last approach feels the cleanest.  It makes it clear you're getting the list type iterable if the object itself is not, otherwise you're getting back the object because it itself is iterable.  Either way, you're getting an Iterable[T] or Iterator[T], as I see it.  It's a little vague in the docstring as to whether the return value is supposed to support the iterator protocol.  It just says "iterable" as the keyword.

### Should I leave the docstrings alone?
Some of the docstrings in utils mention "optional" arguments, but they actually kind of aren't.  When I think 'optional' I admittedly lean more towards `Optional[T]` from the typing module, in that the argument itself could be None, but the docstring's context here is optional in the sense that you don't have to provide that argument (i.e., it has a default value).  I was considering updating the docstrings to change those 'optional' texts to indicate instead that it has a default argument rather than keep the typing ambiguity in there.  This is probably a me thing, since I'm thinking about "or None" when I think "optional", but I wanted an opinion on documentation changes as well.

**Reply by Griatch** [reply]
- I don't mind if code-changes are suggested if you think the code as written makes things harder to read. The Evennia code base represents more than a decade of Python use; we started this back in Python 2.2 days, so it's not too surprising that not everything is as clean is could be :) As you already noted though - this is unrelated to typehints, so such changes should really be in a separate PR. 
- In Google docstring style, when you have a function like this: `def funcname(a, b="foo", c=None, *args, **kwargs)`, you can write it in these two, equivalent forms: 
```
"""
Short description of the function. 

Args: 
    a (str): What `a` does.
    b (str, optional): What b does.
    c (str or None, optional): What c does if given, or left as `None`.
    *args: Other args.
    **kwargs: Other kwargs.
"""
```
Or, alternatively
```
"""
Short description of the function. 

Args: 
    a (str): What a does.
    *args: Other args.

Keyword Args: 
    b (str): What b does.
    c (str or None): What c does if given, or if left as `None`. 
    **kwargs: Other kwargs.
"""
```

These two forms are equivalent and we mix the two styles a bit in the code. "Optional" here does not mean "or None", it tracks if this is a value you have to provide when calling the command or not. If the default value is `None` or something else is irrelevant.

Technically, Google code style allows you to skip the `(type)` parenthesis when providing the type hints as part of the arguments. But I think we should not do this in this case, since the typehints will be separated into a separate typehint package and most people will be reading the doc strings still. Also, from my other comment above - I think it probably makes more sense to not clone the full doc strings to the typehinting stubs; that just sounds like a recipe for having them go out of sync.

**Reply by dvoraen** [reply]
> * I don't mind if code-changes are suggested if you think the code as written makes things harder to read. The Evennia code base represents more than a decade of Python use; we started this back in Python 2.2 days, so it's not too surprising that **not everything is as clean is could be** :) As you already noted though - this is unrelated to typehints, so such changes should really be in a separate PR.

My major concern is deciphering the ambiguity in some areas.  Here is a different example that came up while I was doing `@overload` typing for utils.variable_from_module:

https://github.com/evennia/evennia/blob/dab057f5ccd51c4b1f05b1a5e5504d95206d2a27/evennia/utils/utils.py#L1434-L1457

As written, `if not module` implies that `module` was passed in as an empty string... or None.  It's not *supposed to* be given None according to the docstring, but it *could*, and the code will return `default` in either case, which could be a str or None itself.  My initial thought is that the overload should avoid the "could have this value" cases such as an erroneous None value, but when the code itself handles that case, intended or not, it makes me start wondering as to whether the stub should explicitly handle that case. :)  I'm inclined towards making it erroneous to at least the type checkers, in that None is not an "authorized" value to pass to this function (i.e., it's considered an error/bug for it to be given that).

Does that make sense, insofar as my issue with ambiguity is concerned?

> * In Google docstring style, when you have a function like this: `def funcname(a, b="foo", c=None, *args, **kwargs)`, you can write it in these two, equivalent forms:
*snip*

I had to review that, so I understand your point now wrt the docstrings.  Based on looking at the stubs supplied with Pylance for the standard library, I'm keeping docstrings out of the stubs.  Pylance, at least, will apparently draw from both stubs and actual .py files for the docstrings and typing information, because I get both when the stubs have no docstrings within them.

**Reply by Griatch** [reply]
> As written, if not module implies that module was passed in as an empty string... or None. It's not supposed to be given None according to the docstring, but it could, and the code will return default in either case, which could be a str or None itself. My initial thought is that the overload should avoid the "could have this value" cases such as an erroneous None value, but when the code itself handles that case, intended or not, it makes me start wondering as to whether the stub should explicitly handle that case. :) I'm inclined towards making it erroneous to at least the type checkers, in that None is not an "authorized" value to pass to this function (i.e., it's considered an error/bug for it to be given that).

You are correct in that `None` is technically possible to provide, and would be equivalent to `""`. I think in general, the actual functionality of a function should not be modified by a typing project. So in a case like this, it's the original docstring which is not precise enough - it should really be `(str, module or None)`. So, when there's ambiguity, the code takes precedence over the docstring.

**Comment by dvoraen** [comment]
One question that arose in my tinkering this morning:

In many docstrings, the term "iterable" is used to describe a function argument. As far as I can tell, this seems to really mean "a sequence" in context*, but that's not necessarily true for all iterable objects (not all iterable objects are sequences). What is the intended meaning behind "iterable" in this particular context? My initial impression was to just slap them under the Iterator protocol and call it a day, but more and more I'm starting to think that Sequence or perhaps Collection is more appropriate, as a number of the "iterables" are described as lists and such.

Example docstring, note the description for the `vocabulary` argument:

https://github.com/evennia/evennia/blob/dab057f5ccd51c4b1f05b1a5e5504d95206d2a27/evennia/utils/utils.py#L1674-L1691

\* I haven't found an exception to this statement just yet.  If anything, I found one "iterable" having len() called on it after being checked via is_iter(). :)

Confession time: I took great pleasure at typing lazy_property (particularly `__get__`).  Its type inferencing listing `Any` as the return type drove me NUTS when going through Arx's code. :)

**Reply by Griatch** [reply]
Pythonic style is to not restrict types more than necessary - this is a key function of Python and why one must be careful about typing. In the example given, `vocabulary` could be anything that you can iterate over (list, tuple, even a dict, if you only consider the keys). In practice it's most likely a list or tuple, and I think a 'sequence' is likely a good descriptor.

**Comment by dvoraen** [comment]
I think I'm overthinking and/or "but what if" one aspect of this*, so this post is one part asking for thoughts to help me untangle this mental snarl, one part blog post to explain my thought processes.  The example in question for this post is this function in evennia.utils.utils:

https://github.com/evennia/evennia/blob/eb3fcc0a599feca0124c25d17570cf35dd1a0dbb/evennia/utils/utils.py#L124-L147

As it stands, this is the typing I'm doing for it and other functions like it (that expect ONLY specific values for certain arguments, like `align` here):

```python
def pad(
    text: str, width: int | None = ..., align: Literal["c", "l", "r"] = ..., fillchar: str = ...
) -> str: ...
```

I specifically am choosing `Literal` over `str` and `AnyStr` because the expected values for `align` here (which are explicitly checked for in the code) are the three values in the `Literal` hint.  Anything else causes it to default to "c" in `pad()`'s case.  The issue is that if a _variable_, and not a literal 'c'/'l'/'r' is passed to `pad()`, a type checker won't(?) necessarily know what the value of the argument is, and could flag that as an error.  I don't like the idea of doing an `@overload` because it "could" be a general str (and technically is), but this is where the overthinking comes in.  I feel like I'm concentrating too much on the what-ifs and there needs to be a line drawn somewhere, but every time I try to measure where, the goalpost keeps changing. :)  Part of why I bring it up is that if you're doing type checking on the use of a function, you'd want the analyzer to catch uncertainties with respect to the arguments and not simply say "here's the general overload have fun", no?  In other words, I want to help the analysis tool catch argument errors where it can.  (I'm thinking about this partly from how Pylance would behave; if it can't match one up, it'll provide the generalized typing if it exists.)

I think I need to do more research, particularly on the stdlib overloads.  `open()` is probably where I should look, because I vaguely recall that using `@overload` typing alongside `Literal`.

\* I'm still learning `@overload` nuances here (the 'when and when nots'), which is the source of this post. :)

**Reply by BigJMoney** [reply]
`width: int | None`
Are unions with None preferable to what's more common in other languages:
`width: Optional[None]` ?
I know that pydantic treats them differently, but I don't have enough experience with other packages to know how common that is. My preference would be for the optional over union with None since I think the intent is a little clearer.

*Edit: Python documentation says they're the same actually, so nevermind. I'm not sure why the pydantic thread about this (no pun-intended). Maybe it was different in an older version of Python, or maybe Pydantic reads the type strings directly and changes behavior between optional none and union with none.

**Reply by dvoraen** [reply]
Yeah `Optional[T]` is literally the same as `Union[T, None]`, and 3.9 implemented union syntax to allow for omitting `Optional[T]` and `Union[T1, T2]` entirely in favor of `T | None`.  Personally I kind of like having `Optional[T]` for parameters (only), but consistency is what's going to win out on my end, so `T | None` is what's being used. :)

**Reply by BigJMoney** [reply]
No worries, it's nice and short and easy to get used to.

**Reply by Griatch** [reply]
I think `Literal["c", "l", "r"]` is a good call; it actually adds information to the typehint as well. That said, one can almost go arbitrarily deep with stuff like this; if it adds a lot more complexity and headache, it may be better to just say `: str` and be done with it. 

(and `int | None` is definitely easier to read :+1:  The `Optional/Union` nonsense was horrible.)

**Reply by dvoraen** [reply]
> I think `Literal["c", "l", "r"]` is a good call; it actually adds information to the typehint as well. That said, one can almost go arbitrarily deep with stuff like this; **if it adds a lot more complexity and headache, it may be better to just say `: str` and be done with it**.
> 
> (and `int | None` is definitely easier to read 👍 The `Optional/Union` nonsense was horrible.)

RE: bolded
Definitely.  The point of a type hint is to help the analysis tools, and in the case of Pylance (VSCode) -- and Pycharm, I believe; I don't really use it -- you get Intellisense and code completion benefits, so it's in everyone's interest if the type hints are legible so you know what goes there.  EXCEPTION: decorators, because those can become a tangle quickly.

I went and asked for help from the typing community on the proper way to `@overload` a function, and multiple major players of the typing community (accepted PEP authors) weighed in on my questions, so I have a much less cluttered brain now. :)  If you're curious, it's [here](https://github.com/python/typing/discussions/1291), though I'll warn you that some of the content may result in a "well I _never_" reaction from you, Griatch. :)  (Spoiler alert: inline typing vs. stub library.)

**Comment by dvoraen** [comment]
Today's post is brought to you by the letters F(older) and T(ypes), and the number i.  Yes, the square root of negative one. :)

Our special guest is the Word of the Day: dithering.

### Stuff
In my slow progress through evennia.utils, I came to discover that I'm really in need of defining types (NamedTuples and some Literals in particular) for some specific aspects of the codebase, because they arise in multiple locations of the code.  Here are two examples that feel like they need some unification for one or more components:

#### Supported databases (and their names)

https://github.com/evennia/evennia/blob/62c2558e985dde08c5adcf4fac22b6a2ef5e778f/evennia/utils/utils.py#L1064-L1080

The supported databases could change over time (presumably based on Django's support of them), so I figure that these should be a global constant (tuple? Literal?) of their own, if they're not already*, and functions and methods that utilize the names of those databases should refer to the global holding these database names, rather than be spelled out as in this function docstring.  I'm not sure if I'm taking this a step too far, in that the Literal for the accepted values is self-documenting and creating a type/constant for them may obfuscate those values.  I haven't yet played around with analysis tools in IDEs (Pylance/VSCode, Pycharm) to see how a TypeAlias for a Literal would be displayed to the developer; it's on my todo list.

Anyway, I feel like in typing functions like this, I should be referring to something defined elsewhere instead of "hardtyping" it.  (See below for definition of "hardtyping".)

\* (I haven't gone digging for the global variables for Evennia yet.  As I said, slow progress. :))


#### TickerHandler store_key tuple
https://github.com/evennia/evennia/blob/62c2558e985dde08c5adcf4fac22b6a2ef5e778f/evennia/scripts/tickerhandler.py#L362-L398

This function creates tuples that are used to uniquely identify tickers, which are used here by the TickerHandler as well as by a few functions in evennia.utils.  The docstring is the primary source of information about the tuple returned here, but I feel like it could be made to be more self-documenting.  See below question about creating evennia.typing for my thoughts on this.

#### Comments
I bring these examples up mostly as memos to me, because I'm actually uncertain of how to use GitHub to manage some of my findings for potential code changes. :)  See my first question below for the method to the madness.


### Questions

#### Do forks allow their own issues and project cards to exist in isolation?
~~This is me being ignorant about GitHub.  If I start making issues for code changes in my fork of Evennia, will that let me track these things in my own corner, or will I be cluttering up the evennia/evennia repository with issues I create on my fork?  More specifically, if I open an issue in my fork, does it simultaneously open the same issue in the upstream repository?~~
I have researched this and found my answer.  Setting up my fork to basically be a typing issue tracker to teach myself Issue tracking, Projects, and so on. :)

#### Should evennia.typing exist or should I just inline some of these special types?
The biggest example is the TickerHandler one above.  That is ripe for a NamedTuple, in my estimation, because the tuple is used by multiple parts of the code.  The TickerHandler and the utils module both make use of it, and it seems "unfair" to expect a developer to have to research how an arbitrary tuple works when a NamedTuple will be self-documenting.  I'm thinking that since the TickerHandler is the primary creator and user of this tuple, any type creation should go there.  However, if other aspects of the code use it, perhaps "types made to explain code" should exist in evennia.typing as a catch-all location for where to find them, as creating a type such as a NamedTuple isn't itself essential to the operation of the code.

My point is: putting a title on the book doesn't change the contents, but it makes it much easier to find on the shelf. :)

There are a couple things I think should be generalized in an evennia.typing module.  The supported databases, for one, as well as possibly the justification codes used by `utils.pad()` and `utils.justify()`.  The dithering is me debating about whether this should carry the caveat of code changes to emphasize validation of values, or if we should just let the `Literal`s do all the work here.  I'm inclined to just letting the `Literal` values do all the work, myself.  (Memo to me: reread the Literal PEP.)


### Definition of Hardtyping
I'm coining this term to mean "spelling out a type directly in the code using it."  The best way I can explain that is the following comparison and contrast:

```python
def foo(input: Literal["hi", "hello", "greetings") -> str:
    return input

GreetingString: TypeAlias = Literal["hi", "hello", "greetings"]

def bar(input: GreetingString) -> str:
    return input
```
In comparing `foo()` and `bar()` here, `foo()` would be considered "hardtyped"; the `Literal` is directly injected into the parameter list and spelled out there.  Unlike `bar()`, any change to `foo()`'s "acceptable" input would require going to `foo()` and updating the `Literal` directly, in addition to any other code that uses the same values.  With the `TypeAlias`, though, we can just modify that, and never have to touch `bar()` and other code that uses `GreetingString`.

**Comment by Griatch** [comment]
> The supported databases could change over time (presumably based on Django's support of them), so I figure that these should be a global constant (tuple? Literal?) of their own, if they're not already*, and functions and methods that utilize the names of those databases should refer to the global holding these database names, rather than be spelled out as in this function docstring. I'm not sure if I'm taking this a step too far, in that the Literal for the accepted values is self-documenting and creating a type/constant for them may obfuscate those values. I haven't yet played around with analysis tools in IDEs (Pylance/VSCode, Pycharm) to see how a TypeAlias for a Literal would be displayed to the developer; it's on my todo list.

While maybe an interesting experiment in learning typing, I think this is taking it too far. By restricting literals we are starting to affect more than types, but also the code content, which goes beyond the purview of typing.

> Do forks allow their own issues and project cards to exist in isolation?
You already figured this out - yes, your fork is its own thing, you can have your own project cards and issues. 

> Should evennia.typing exist or should I just inline some of these special types?

For the example; the tuple is really acting as a static key. From the outside it's not something the user ever modifies or is expected to know the contents of; we could maybe have hashed it into a string too. You are expected to get it out, store it and then use it whenever you need to operate on a given ticker. It sounds like this is something that can be defined along with the rest of the stub.

> Definition of Hardtyping

While I like the look of `foo` better, I think it's going too far to enforce literal input choicers like this. Not only does it put an immense overhead on development (and on making the types too), it also goes beyond what a 'type' is, moving into code content. 

```python
def foo(input: str) -> str: 
    return input
```
is all that is needed here.
